### DISCUSSION TITLE
Memory model as type.

### DISCUSSION THREAD

**Discussion by YagaoDirac** [hsutter/cppfront#865] [Suggestions]
Hi.
I briefly read the titles of all the discussions, I don't know if this suggestion overlaps any of them. Sorry for that if it happens.

The idea:
Memory model can help compiler to do some checking, just like the type system does. So can we also add the memory model to some where in the type system.

I mean, all the unique_ptr, share_ptr, in, inout, instant value, left value, right value, in my mind are all memory models. In cpp1, smart pointers are implemented in std lib, but a lot others are made in syntax. I'm confused sometimes when they are implicitely converted into some others. If they are part of the type system, at least I can cout them.

So, my idea is like, the type system should become 2 layers. 1 is the the same as it is in cpp1 now(int, bool, float, object), the 2nd is the memory model. So, we would have:
```
int a = 42; //a is int/left value, 42 is int/instant value.
```
```
some_class b = {}; // b is some_class/unique_ptr, {} is something...
some_class<share_ptr> c = {}; // c is some_class/share_ptr.
```
```
int <share_ptr> d = 42; // d is int//share_ptr. If you need to store an int in reference counting way, this is how you do it.
```
```
func(in param_in:int, inout param_inout:float)->something = {something else here.} //param_in is int/instant value, param_inout is float/some kind of reference type.
```

Optional. 
Nullable can also be part of the type system. Maybe this is already decided.
Generic and constraint(for better error info) in memory model. Since the "last use"check is already there, so it's possible to seperate the && way from others, with a much better error info.
The restrict keyword from c. Maybe it's possible to pick this up this time. Or do we need it anymore?
Maybe this is also a better way to introduce pmr in cpp17 into cpp2.

**Comment by JohelEGP** [comment]
> Optional.
> Nullable can also be part of the type system. Maybe this is already decided.

Yes, see <https://github.com/hsutter/cppfront/issues/27#issuecomment-1264514145>.

**Comment by JohelEGP** [comment]
>  Maybe this is also a better way to introduce pmr in cpp17 into cpp2.

You can already use the Cpp1 standard library in Cpp2.

**Reply by YagaoDirac** [reply]
Cool!

**Reply by JohelEGP** [reply]
See also <https://github.com/hsutter/cppfront/discussions/620#discussioncomment-7745690>.

**Reply by YagaoDirac** [reply]
Thanks. Hsutter also replied in that thread.

**Comment by MaxSagebaum** [comment]
What would be the benefit of writing i: int <shared_ptr> instead of writing i: shared_ptr<i>.

Also your angled brackets would clash with template definitions. What other symbol would you propose to eliminate the clash?

**Reply by YagaoDirac** [reply]
I dont propose syntax. It's the team's job.

**Comment by hsutter** [comment]
> all the unique_ptr, share_ptr, in, inout, instant value, left value, right value, in my mind are all memory models.

Thanks for the feedback! I've thought about this and I don't think it's a direction I want to pursue for now. I don't understand well what existing known pitfalls this aims to solve. And I think it's an important feature of C++ that I can allocate an object in any location and not have its location be part of its type (including that I can invoke `f(T)` with any `T` object, whether its lifetime happens to be static or controlled by a smart pointer or otherwise.

**Reply by YagaoDirac** [reply]
Hi.
I'm not pro in making programming lang. 
My idea is mainly for, if it's a type, then we can do everything we do to type to it. We can print out the memory model, convert the memory model in a verbose way.
It's more about ***learnability***, not read ability or writability. 
Actually I found the hardest part is learning. I'm learnt 7 programming languages, glanced 2 or 3 more, 1.5 natural lang other than my first lang, and a lot. The era really forces us to learn as much. 
I'm recently having some fun on a personal and "only for fun" language. I tried to shrink the amount of learning to extreme, while maintain other good things as possible. It looks a bit weird, but I really think it's the way. It's very easy to learn especially for people who already have at least some exp in programming. It's not uploaded now. If you are interested in it, I'll prioritize it a bit so to upload it as soon.

**Comment by SebastianTroy** [comment]
Are you suggesting a debug meta type that printed to console whenever a type is constructed/destroyed/copied/moved?


On 22 December 2023 04:25:47 YagaoDirac ***@***.***> wrote:

Hi.
I'm not pro in making programming lang.
My idea is mainly for, if it's a type, then we can do everything we do to type to it. We can print out the memory model, convert the memory model in a verbose way.
It's more about learnability, not read ability or writability.
Actually I found the hardest part is learning. I'm learnt 7 programming languages, glanced 2 or 3 more, 1.5 natural lang other than my first lang, and a lot. The era really forces us to learn as much.
I'm recently having some fun on a personal and "only for fun" language. I tried to shrink the amount of learning to extreme, while maintain other good things as possible. It looks a bit weird, but I really think it's the way. It's very easy to learn especially for people who already have at least some exp in programming. It's not uploaded now. If you are interested in it, I'll prioritize it a bit so to upload it as soon.

**Reply by YagaoDirac** [reply]
Yeah, maybe your way is better. If it's limited in debug, it may be good enough. And it doesn't add up to in code features.
I want to see it somehow, so that I have to deal with less guess work. It's really a disaster to learn something new while cannot see it directly. Also, it maybe clearer to control the convertion if it's denoted explicitly.
